drivers/i2c/sun6i-a31-i2c.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <clock.h>
+#include <dm.h>
+#include <error.h>
+#include <i2c.h>
+#include <mmio.h>
+#include <stddef.h>
+#include <util.h>
+
+#define I2C_CTRL_REG 0x0c
+#define I2C_DATA_REG 0x08
+
+enum I2CStatus {
+	START_COND_TX        = 0x08,
+	START_COND_TX_REPEAT = 0x10,
+	ADDR_WRITE_RX_ACK    = 0x18,
+	ADDR_WRITE_RX_NACK   = 0x20,
+	DATA_TX_ACK          = 0x28,
+	DATA_TX_NACK         = 0x30,
+	ADDR_READ_TX_ACK     = 0x40,
+	ADDR_READ_RX_ACK     = 0x48,
+	DATA_RX_NACK         = 0x58,
+	IDLE                 = 0xf8,
+};
+
+static int
+sun6i_a31_i2c_read(struct device *dev, uint8_t *data)
+{
+	/* Wait for data. */
+	if (!i2c_status_verify(dev, DATA_RX_NACK))
+		return ENODEV;
+
+	/* Read data. */
+	*data = mmio_read32(dev->regs + I2C_DATA_REG) & BITMASK(0, 7);
+
+	/* Move to next state. */
+	mmio_setbits32(dev->regs + I2C_CTRL_REG, BIT(3));
+
+	return SUCCESS;
+}
+
+static int
+sun6i_a31_i2c_start(struct device *dev, int i2c_op)
+{
+	/* Send start. */
+	mmio_setbits32(dev->regs + I2C_CTRL_REG, BIT(5) | BIT(3));
+
+	/**
+	 * TODO(@drewwalters96): This is not true on every restart.
+	 */
+	if (i2c_op == I2C_WRITE) {
+		if (!i2c_status_verify(dev, START_COND_TX))
+			return EIO;
+	} else {
+		if (!i2c_status_verify(dev, START_COND_TX_REPEAT))
+			return EIO;
+	}
+
+	return SUCCESS;
+}
+
+static int
+sun6i_a31_i2c_stop(struct device *dev)
+{
+	/* Send stop. */
+	mmio_clearbits32(dev->regs + I2C_CTRL_REG, BIT(4) | BIT(3));
+
+	if (!i2c_status_verify(dev, IDLE))
+		return EIO;
+
+	return SUCCESS;
+}
+
+static int
+sun6i_a31_i2c_write(struct device *dev, uint8_t data)
+{
+	/* Write data. */
+	mmio_setbits32(dev->regs + I2C_DATA_REG, data);
+
+	if (!i2c_status_verify(dev, DATA_TX_ACK))
+		return ENODEV;
+
+	/* Move to next state. */
+	mmio_setbits32(dev->regs + I2C_CTRL_REG, BIT(3));
```

Comment:
This here is asking the hardware to actually send the data. So we have to do it before waiting for the "data has been sent and ACKed" status